### ISSUE TITLE
How to programmatically run XUnit tests (for profiling with VS2015)?

### ISSUE THREAD

**Issue by gsscoder** [xunit/xunit#542]
I'd like to add a console application that run tests in my test project for **profiling it with VS2015**.

I've scrutinized console runner [source](https://github.com/xunit/xunit/blob/master/src/xunit.console/Program.cs), but since I don't like (and don't want) to _re-invent the wheel_, I'm asking here the minimal API calls to execute tests from a referenced DLL.

At least I'd like to know how to run a single test (I can arrange discovery with reflection or manually decide what run):
```csharp
var myTestClass = /* get reference of test class */
ExecuteXUnitTests(myTestClass); /* ??? */
```

How can I do that without re-creating a custom XUnit console runner?

**Comment by bradwilson** [comment]
Here are the steps:

- Bring in the [xunit.runner.utility](https://www.nuget.org/packages/xunit.runner.utility) NuGet package
- Create an instance of [XunitFrontController](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/xunit.runner.utility/Frameworks/XunitFrontController.cs)
- Create an instance of [TestDiscoveryVisitor](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/common/TestDiscoveryVisitor.cs)
- Call [Find](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/abstractions/xunit.abstractions/Frameworks/ITestFrameworkDiscoverer.cs#L28-L35) on the front controller, passing in the TestDiscoveryVisitor for the `discoveryMessageSink`.
- Wait on the [Finished](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/common/TestMessageVisitor.cs#L512) event on the TestDiscoveryVisitor
- Decide which of the test case(s) you got back that you want to run
- Create a class which derives from [TestMessageVisitor&lt;ITestAssemblyFinished&gt;](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/common/TestMessageVisitor.cs#L498-L499), and override the methods that responds to messages you care about. A good first example is [XmlTestExecutionVisitor](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/common/XmlTestExecutionVisitor.cs), which watches the execution message to put together xUnit.net's XML output.
- Call [RunTests](https://github.com/xunit/xunit/blob/9d22ca78d33fb375d0b9002cae76462ddbc91821/src/abstractions/xunit.abstractions/Frameworks/ITestFrameworkExecutor.cs#L26-L32) on the front controller, passing your execution visitor as the `executionMessageSink`.
- Wait on the Finished event on your execution visitor to know when the execution is complete.

I've opened an issue to make a meta-runner class which makes this easier.

**Comment by bradwilson** [comment]
@gsscoder Check out this sample and tell me what you think:

https://github.com/xunit/samples.xunit/blob/master/TestRunner/Program.cs

**Comment by bradwilson** [comment]
I'm also soliciting general comment on which style people would prefer:

**Delegate style**

```csharp
public delegate void ExecutionCompleteHandler(int totalTests, int testsFailed, int testsSkipped, decimal executionTime);

public ExecutionCompleteHandler OnExecutionComplete { get; set; }
```

**DTO style**

```csharp
public class ExecutionCompleteData
{
    public int TotalTests { get; set; }
    public int TestsFailed { get; set; }
    public int TestSkipped { get; set; }
    public decimal ExecutionTime { get; set; }
}

public Action<ExecutionCompleteData> OnExecutionComplete { get; set; }
```

**Comment by sapiens** [comment]
DTO style

**Comment by cmatskas** [comment]
DTO gets my vote

**Comment by myefforts** [comment]
+1 for DTO

**Comment by gsscoder** [comment]
@bradwilson, wow, thanks for all useful informations! :)

I'll put myself on work ASAP and tell you.

For now, again, thank you!

**Comment by gsscoder** [comment]
@bradwilson, the AssemblyRunner (of which provided source) should be present in xunit.runners.utility package?

**Comment by gsscoder** [comment]
@bradwilson, I've manually added AssembyRunner and its source dependencies.

When I run it in visual studio (`config=Debug`), if a test correctly throwns an exception execution stops. I think this a normal behaviour, right?

The code is here for your review: https://github.com/gsscoder/commandline/tree/addprof/tests/CommandLine.Tests.Runner. (Temp dev branch: `addprof`).

I'll try to profile it, now... :)

**Comment by gsscoder** [comment]
@bradwilson, the raised exception prevents profile from collecting data... :(

Any suggestion?

(Thanks for all help provided until now).

**Comment by bradwilson** [comment]
@gsscoder writes:
> the AssemblyRunner (of which you provided source link) should be present in xunit.runners.utility package?

It's in the CI builds available on https://www.myget.org/F/xunit (it has not shipped officially yet).

> the raised exception prevents profile from collecting data... :(

I guess I'm not clear what it is you're trying to do. Can you describe what you're trying to accomplish, and what steps you're taking?

**Comment by bradwilson** [comment]
I've [switched to DTO style](https://github.com/xunit/xunit/commit/1d92bb4ddcb3995a25828034ce56baf79a6fe2ee) and updated [the sample runner](https://github.com/xunit/samples.xunit/blob/3f55e554e7de7eb2cd9802fa9e73a706520646cd/TestRunner/Program.cs).

/cc @gsscoder @sapiens @cmatskas @myefforts

**Comment by gsscoder** [comment]
:+1: for DTO style.

I can close the issue since the problem in collecting performance data is not up to XUnit project.

Thanks for excellent support!

**Comment by duongphuhiep** [comment]
I cannot get it working:

http://stackoverflow.com/questions/42764789/how-to-make-a-xunit-test-runner-programaticly-in-the-application

```
using Xunit.Runners;

var r = AssemblyRunner.WithAppDomain("WebServices.dll");
var e = new ManualResetEvent(false);
r.OnTestFinished = i =>
{
  e.Set();
  Log.Info("OnTestFinished");

};
r.OnTestStarting = i =>
{
  Log.Info("OnTestStarting");
};
r.Start();
e.WaitOne();
```

**Comment by vmachacek** [comment]
I manage to run the tests, but when the runner finishes it wont invoke the `TestAssemblyExecutionFinishedEvent` or `TestExecutionSummaryEvent` I'm probably missing something simple. I made sample repo here if someone fancy a peek 
https://github.com/vmachacek/xunit-run-test/blob/master/TestRunner/Program.cs

**Comment by vmachacek** [comment]
ExecutionEventSink solved the problem for me. This is the final code

https://pastebin.com/q9UD87mR

**Comment by holbizmetrics** [comment]
I managed to do all as told, but I have no idea how to make ITestOutputHelper work.
How could this be done then?

My distinct tries:

1.) I tried to create a constructor and do it with reflection.
With all variants imaginable.
public class AllTestsTest
{
    private ITestOutputHelper outputHelper { get; }
    public AllTestsTest(IOutputHelper outputHelper)
    {
        this.outputHelper = outputHelper; 
    }

    [Fact] 
    public void Test()
    {    
       ITestOutputHelper outputHelper = new TestOutputHelper();
       outputHelper.WriteLine("Test Output");
    }
}


2.) I also tried to create a test where I did the following in the test:

[Fact]
public void Test()
{   
   ITestOutputHelper outputHelper = new TestOutputHelper();
   outputHelper.WriteLine("Test Output");
}

Do I need to register something for TestOutputHelper?
The messageBus and the ITest? But how to get it?

Thanks for any help in advance.
